### DISCUSSION TITLE
Divide a c# script in several files

### DISCUSSION THREAD

**Discussion by rvgfox** [TabularEditor/TabularEditor3#1210] [🙋♀️TE3 Q&A] [answered]
I've a script with a lof off lines, but really it can divide in more small blocks.

My question it's if each block can be store in a file.

I'm running the c# script in TE2 command line.

**Comment by rvgfox** [comment]
Hi @otykier can you help me?

**Reply by otykier** [reply]
There are a few options available today, but this is still an area we’re working to improve.

The simplest option is to store individual blocks as macros. You can then invoke them from other scripts/macros, by calling the [CustomAction](https://docs.tabulareditor.com/api/TabularEditor.Shared.Scripting.ScriptHost.html#TabularEditor_Shared_Scripting_ScriptHost_CustomAction_System_String_) method. The downside of this approach is that you can’t easily pass variables/objects between scripts/macros code.

A more advanced option is to create your own .NET library using Visual Studio. For TE2 you would need to target .NET Framework 4.7.2. For TE3 it’s .NET 6. You can then [reference the external library from within your scripts/macros](https://docs.tabulareditor.com/te2/Advanced-Scripting.html#net-references).

**Comment by rvgfox** [comment]
Thanks @otykier The CustomAction method sounds easy to use. Is it possible to use with TE2 in a main script called by a command line?

**Reply by otykier** [reply, marked as answer]
Yes, provided the macros are also available in TE2.

**Comment by rvgfox** [comment]
Hi @otykier I have changed my script to use CustomActions, adding it as macros, and in TE2 it works but not in the bach using command line.
The original script works in command line without problems but the new one using customAction method does not work.

any inea?

Thanks in advance.

Ramon

**Reply by otykier** [reply]
Are you running the command line on the same machine where you originally saved the macros? If not, you will need to copy the `%LocalAppData%\MacroActions.json` file (which contains the definition of the macros), over to the machine where you’re using the command line.

**Reply by rvgfox** [reply]
Yes, I have installed the macros on the same machine, and running directly with TE2 without errors but not on the command line.

Previously, with the "full script" (without CustomActions) it worked from the command line perfectly.

I am always using the same command line in both cases.

**Reply by otykier** [reply]
I'm assuming you're using the latest version of TE2? I tested this just now, and I'm not seeing any issues:

<img width="864" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/8976200/c6aabbba-3be8-40b5-9a56-b068802e8066">

In this case, I saved a macro with the name "SayHello", with the code:

```csharp
Info(string.Format("Hello. You have selected {0} objects!", Selected.Count));
```

Please be aware that the macro must support the "Model" context, since there is no current selection of objects when running a script through the command line. So in other words, when you save the macro, make sure that at least the "Model" context has been checked:

<img width="502" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/8976200/a6504851-cbbd-40c7-a241-73a16e0f0ce4">

If you need to pass in a collection of objects, such that the `Selected` object inside the macro isn't empty, you should call `CustomAction` like so from the main script:

```csharp
CustomAction(Model.Tables, "SayHello");
```

<img width="961" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/8976200/7ab03582-f8ac-4c46-8c13-92711dbe1595">

**Reply by rvgfox** [reply]
Thanks @otykier I tried with the command line directly and it works.
The problem it's that the proccess runs in an unattended job using a service with the logon as "Local system account" and the CustomActions cannot find them.
How can I solve it?

**Reply by rvgfox** [reply]
@otykier I tried to put the file MacroActions.jso in the same folder than the main script without result.

**Reply by otykier** [reply]
Apologies for the late reply. Take a look at [this Stackoverflow answer](https://serverfault.com/a/749419). Apparently, `%systemroot%\ServiceProfiles\LocalService` would be the root of the local system account on Windows 10 (and also Windows 11 - I just tested this).

So that means you should put the MacroActions.json file under:

```
%systemroot%\ServiceProfiles\LocalService\AppData\Local\TabularEditor
```

(create the folder if it doesn't already exist)
